Principles/6 - Declarative over imperative.md

```diff
@@ -0,0 +1,27 @@
+# Declarative over imperative
+
+Imperative code is often a natural choice for developers who have spent a lot of time in C-style languages. The primitives of many popular programming languages are very easy to get started with writing highly imperative code. However, building imperative APIs can lead to much more error prone systems, more difficult to read code, and it being harder to discern the intention from the implementation for a given piece of logic.
+
+We prefer to write code that exposes a declarative API than an imperative API. A declarative API will let you define what results you would like to see, rather than the way you want them to happen. This type of API leads to simpler code with a higher degree of readability, maintainability, and composability. Where imperative code must exist we often counsel using more declarative and immutable styles within it.
+
+Below are a few examples of decisions that demonstrate this principle, some at the highest level of technology choice, and some at the smallest level of individual function implementations:
+
+* We prefer foundational technologies that are declarative, such as JSX and GraphQL, rather than imperative DOM manipulation and REST calls. On the build tooling side we choose the declarative Webpack over more imperative tools like Gulp.
```

Comment:
I would separate some of these out, we chose React over other librar/ technologyy X(web components maybe?) because of the declarative API. We chose GraphQL over REST.

Reply:
The build tool example is not a particularly strong one so I would ditch it